hbase-server/src/main/java/org/apache/hadoop/hbase/master/ServerManager.java

```diff
@@ -227,16 +264,47 @@ ServerName regionServerStartup(RegionServerStartupRequest request, int versionNu
     final String hostname =
       request.hasUseThisHostnameInstead() ? request.getUseThisHostnameInstead() : isaHostName;
     ServerName sn = ServerName.valueOf(hostname, request.getPort(), request.getServerStartCode());
+
+    // Check if the HMaster is configured to reject decommissioned hosts or not.
+    // When configured to do so, all hosts of decommissioned RegionServers are prevented from
+    // reporting for duty; otherwise, we accept them unless they come back with an identical
+    // ServerName that we can find in the decommissioned hosts.
+    // See HBASE-28342 for details.
+    if (shouldHostBeRejected(hostname)) {
+      LOG.warn("Rejecting RegionServer {} from reporting for duty because Master is configured "
+        + "to reject decommissioned hosts and this host was marked as such in the past.", sn);
+
+      throw new HostIsConsideredDecommissionedException(
+        "Master is configured to reject decommissioned hosts");
+    }
+
     checkClockSkew(sn, request.getServerCurrentTime());
     checkIsDead(sn, "STARTUP");
     if (!checkAndRecordNewServer(sn, ServerMetricsBuilder.of(sn, versionNumber, version))) {
-      LOG.warn(
-        "THIS SHOULD NOT HAPPEN, RegionServerStartup" + " could not record the server: " + sn);
+      LOG.warn("THIS SHOULD NOT HAPPEN, RegionServerStartup could not record the server: {}", sn);
     }
     storage.started(sn);
     return sn;
   }
 
+  /**
+   * Checks if the server is configured to reject decommissioned hosts, if that's the case, it then
+   * checks if the host exists in the list of drained servers, and when it's there it returns true;
+   * otherwise returns false.
+   * @param hostname The hostname of the server
+   */
+  private boolean shouldHostBeRejected(String hostname) {
```

Comment:
Why not follow the existing style? Wrap up this logic all inside of a `checkDecommissionedStatus(sn)` call, like we do for the other invariants.

Reply:
Do you mean wrap the whole logic in this method + what's also added in the `regionServerStartup` method? I didn't think much of the style but now that I pay attention to it, I like it. Will fix in a bit.